core/src/main/java/org/pentaho/di/core/logging/MetricsRegistry.java

```diff
@@ -0,0 +1,122 @@
+
+
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import org.pentaho.di.core.metrics.MetricsSnapshotInterface;
+
+/**
+ * This singleton will capture all the metrics coming from the various log channels based on the log channel ID.
+ *
+ * @author matt
+ *
+ */
+public class MetricsRegistry {
+  private static MetricsRegistry registry = new MetricsRegistry();
+
+  private Map<String, Map<String, MetricsSnapshotInterface>> snapshotMaps;
+  private Map<String, Queue<MetricsSnapshotInterface>> snapshotLists;
+
+  public static MetricsRegistry getInstance() {
+    return registry;
+  }
+
+  private MetricsRegistry() {
+    snapshotMaps = new ConcurrentHashMap<String, Map<String, MetricsSnapshotInterface>>();
+    snapshotLists = new ConcurrentHashMap<String, Queue<MetricsSnapshotInterface>>();
+  }
+
+  public void addSnapshot( LogChannelInterface logChannel, MetricsSnapshotInterface snapshot ) {
+    MetricsInterface metric = snapshot.getMetric();
+    String channelId = logChannel.getLogChannelId();
+    switch ( metric.getType() ) {
+      case START:
+      case STOP:
+        Queue<MetricsSnapshotInterface> list = getSnapshotList( channelId );
+        list.add( snapshot );
+
+        break;
+      case MIN:
+      case MAX:
+      case SUM:
+      case COUNT:
+        Map<String, MetricsSnapshotInterface> map = getSnapshotMap( channelId );
+        map.put( snapshot.getKey(), snapshot );
+
+        break;
+      default:
+        break;
+    }
+  }
+
+  public Map<String, Queue<MetricsSnapshotInterface>> getSnapshotLists() {
+    return snapshotLists;
+  }
+
+  public Map<String, Map<String, MetricsSnapshotInterface>> getSnapshotMaps() {
+    return snapshotMaps;
+  }
+
+  /**
+   * Get the snapshot list for the given log channel ID. If no list is available, one is created (and stored).
+   *
+   * @param logChannelId
+   *          The log channel to use.
+   * @return an existing or a new metrics snapshot list.
+   */
+  public Queue<MetricsSnapshotInterface> getSnapshotList( String logChannelId ) {
+    Queue<MetricsSnapshotInterface> list = snapshotLists.get( logChannelId );
+    if ( list == null ) {
+      list = new ConcurrentLinkedQueue<MetricsSnapshotInterface>();
+      snapshotLists.put( logChannelId, list );
+    }
+    return list;
+
+  }
+
+  /**
+   * Get the snapshot map for the given log channel ID. If no map is available, one is created (and stored).
+   *
+   * @param logChannelId
+   *          The log channel to use.
+   * @return an existing or a new metrics snapshot map.
+   */
+  public Map<String, MetricsSnapshotInterface> getSnapshotMap( String logChannelId ) {
+    Map<String, MetricsSnapshotInterface> map = snapshotMaps.get( logChannelId );
+    if ( map == null ) {
+      map = new ConcurrentHashMap<String, MetricsSnapshotInterface>();
+      snapshotMaps.put( logChannelId, map );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **110, 111, and 113**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and returns that reference to its caller at line: **115**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.